nucliadb/nucliadb/ingest/orm/entities.py

```diff
@@ -0,0 +1,261 @@
+            deleted.update(deg.entities_groups)
+        return deleted
+
+    async def get_entitiesgroup_inner(self, group: str) -> Optional[EntitiesGroup]:
+        stored_entities = await self.get_stored_entities(group)
+        indexed_entities = await self.get_indexed_entities(group)
+        entities = self.merge_entities(stored_entities, indexed_entities)
+        return entities
+
+    async def get_stored_entities(self, group: str) -> Optional[EntitiesGroup]:
+        key = KB_ENTITIES_GROUP.format(kbid=self.kbid, id=group)
+        payload = await self.txn.get(key)
+        if payload is None:
+            return None
+
+        egd = EntitiesGroup()
+        egd.ParseFromString(payload)
+        return egd
+
+    async def get_indexed_entities(self, group: str) -> List[str]:
+        node, shard_id = random.choice(
+            [node async for node in self.kb.iterate_kb_nodes()]
+        )
+        request = RelationSearchRequest(
+            shard_id=shard_id,
+            prefix=RelationPrefixSearchRequest(
+                prefix="",
+                node_filters=[
+                    RelationNodeFilter(
+                        node_type=RelationNode.NodeType.ENTITY, node_subtype=group
+                    )
+                ],
+            ),
+        )
+        results = await node.reader.RelationSearch(request)
+
+        entities = [node.value for node in results.prefix.nodes]
+        return entities
+
+    @staticmethod
+    def merge_entities(
+        stored: Optional[EntitiesGroup], indexed: List[str]
+    ) -> Optional[EntitiesGroup]:
+        if not indexed:
+            return stored
+
+        title = ""
+        color = ""
+        custom = False
+        entities = {}
+
+        if stored is None:
+            custom = False
+
+            for entity in indexed:
+                entities[entity] = Entity(
+                    value=entity,
+                    status=Entity.DiffStatus.NORMAL,
+                )
+        else:
+            title = stored.title
+            color = stored.color
+            custom = stored.custom
+
+            for name, entity in stored.entities.items():
+                # skip entities marked as deleted, as we provide a custom view
+                if entity.status != Entity.DiffStatus.DELETED:
+                    entities[name] = entity
+
+            for name in indexed:
+                if (
+                    name in stored.entities
+                    and stored.entities[name].status == Entity.DiffStatus.DELETED
+                ):
+                    continue
+
+                entities.setdefault(
+                    name, Entity(value=name, status=Entity.DiffStatus.NORMAL)
+                )
+
+        eg = EntitiesGroup(entities=entities, title=title, color=color, custom=custom)
+        return eg
+
+    async def index_entities(self, group: str, entities: EntitiesGroup):
+        # TODO properly indexing of SYNONYM relations
+        nodes = {
```

Comment:
If I am not wrong, the `nodes` term is being used twice in this function with different meaning: one is for nodes of the graph and the one below for nodes of nucliadb.

What about `graph_nodes` to disambiguate?

Reply:
Completely agree